File: github-token-is-read-only-by-default/node/actions-permissions-audit.mjs
/**
 * Audit GITHUB_TOKEN permissions: too little to work, or more than needed.
 *
 * The 403 message never names the missing scope, so this maps the operation a job
 * performs back to the scope it requires.
 */
import { readdir, readFile } from 'node:fs/promises';
import path from 'node:path';

const API = 'https://api.github.com';

// Operations that write, and the scope each one needs. The API will not tell you.
const WRITE_HINTS = [
  [/\bgit\s+push\b|actions\/create-release|softprops\/action-gh-release/, 'contents: write'],
  [/gh\s+release\s+create|\bgh\s+api\b.*releases/, 'contents: write'],
  [/gh\s+pr\s+comment|actions\/github-script.*createComment/, 'pull-requests: write'],
  [/docker\/build-push-action|npm\s+publish|gh\s+api.*packages/, 'packages: write'],
  [/aws-actions\/configure-aws-credentials|id-token/, 'id-token: write'],
];

/** Pure decision function: which scopes does this workflow appear to need? */
export function neededScopes(text) {
  return [...new Set(WRITE_HINTS.filter(([re]) => re.test(text)).map(([, s]) => s))].sort();
}

export function declaredScopes(text) {
  return [...new Set([...text.matchAll(/^\s*([a-z-]+):\s*write\s*$/gm)].map((m) => m[1]))].sort();
}

export function gaps(text) {
  const have = new Set(declaredScopes(text));
  return neededScopes(text).filter((s) => !have.has(s.split(':')[0].trim()));
}

async function main() {
  const repo = process.argv[process.argv.indexOf('--repo') + 1];
  const dir = '.github/workflows';
  const token = (process.env.GITHUB_TOKEN || "dummy-github-token");
  if (!token) { console.error('set GITHUB_TOKEN'); process.exit(2); }

  const res = await fetch(`${API}/repos/${repo}/actions/permissions/workflow`, {
    headers: { Authorization: `Bearer ${token}`, Accept: 'application/vnd.github+json' },
  });
  if (res.ok) {
    const { default_workflow_permissions: d } = await res.json();
    if (d === 'write') console.warn(`${repo} defaults to WRITE: every workflow token can push`);
    else console.log(`${repo} defaults to ${d}`);
  }

  let failed = false;
  for (const f of (await readdir(dir)).filter((n) => /\.ya?ml$/.test(n))) {
    const text = await readFile(path.join(dir, f), 'utf8');
    const missing = gaps(text);
    if (missing.length) { failed = true; console.error(`${f} needs ${missing.join(', ')}`); }
    else console.log(`${f}: permissions look sufficient`);
  }
  if (failed) {
    console.error('a missing scope surfaces as: 403 Resource not accessible by integration');
  }
  process.exit(failed ? 1 : 0);
}

if (import.meta.url === `file://${process.argv[1]}`) main();

File: github-token-is-read-only-by-default/python/actions_permissions_audit.py
"""Audit GITHUB_TOKEN permissions: too little to work, or more than needed.

The 403 message never names the missing scope, so this maps the operation a job
performs back to the scope it requires. It also flags repositories still defaulting
to write permissions, which is the same problem pointing the other way.
"""
import argparse
import logging
import os
import re
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("actions_permissions_audit")

API = "https://api.github.com"

# Operations that write, and the scope each one needs. The API will not tell you.
WRITE_HINTS = [
    (re.compile(r"\bgit\s+push\b|actions/create-release|softprops/action-gh-release"),
     "contents: write"),
    (re.compile(r"gh\s+release\s+create|\bgh\s+api\b.*releases"), "contents: write"),
    (re.compile(r"gh\s+pr\s+comment|actions/github-script.*createComment"),
     "pull-requests: write"),
    (re.compile(r"docker/build-push-action|npm\s+publish|gh\s+api.*packages"),
     "packages: write"),
    (re.compile(r"aws-actions/configure-aws-credentials|id-token"), "id-token: write"),
]


def needed_scopes(workflow_text):
    """Pure decision function: which scopes does this workflow appear to need?"""
    return sorted({scope for pattern, scope in WRITE_HINTS
                   if pattern.search(workflow_text)})


def declared_scopes(workflow_text):
    """Scopes the workflow actually grants, anywhere in the file."""
    return sorted(set(re.findall(r"^\s*([a-z-]+):\s*write\s*$", workflow_text, re.M)))


def gaps(workflow_text):
    """What the workflow needs but has not granted."""
    have = {s.split(":")[0] for s in declared_scopes(workflow_text)}
    return [s for s in needed_scopes(workflow_text) if s.split(":")[0].strip() not in have]


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--repo", required=True)
    ap.add_argument("--workflow-dir", default=".github/workflows")
    args = ap.parse_args()

    token = os.environ.get("GITHUB_TOKEN")
    if not token:
        log.error("set GITHUB_TOKEN")
        return 2
    headers = {"Authorization": f"Bearer {token}",
               "Accept": "application/vnd.github+json"}

    r = requests.get(f"{API}/repos/{args.repo}/actions/permissions/workflow",
                     headers=headers, timeout=30)
    if r.ok:
        default = r.json().get("default_workflow_permissions")
        if default == "write":
            log.warning("%s defaults to WRITE: every workflow token can push, release "
                        "and comment whether it needs to or not", args.repo)
        else:
            log.info("%s defaults to %s", args.repo, default)

    from pathlib import Path
    failed = False
    for wf in sorted(Path(args.workflow_dir).glob("*.y*ml")):
        text = wf.read_text(encoding="utf-8")
        missing = gaps(text)
        if missing:
            failed = True
            log.error("%s needs %s but does not declare it",
                      wf.name, ", ".join(missing))
        else:
            log.info("%s: permissions look sufficient", wf.name)
    if failed:
        log.error("a missing scope surfaces as: 403 Resource not accessible by "
                  "integration -- the message never says which one")
    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(main())
